### ISSUE TITLE
Whole computer freezes when it gets to StructureFromMotionCL

### ISSUE THREAD

**Issue by ThePinkUnicorn6** [openphotogrammetry/meshroomcl#4]
When it completes CameraInit and strts processing StructureFromMotionCL, the computer freezes. The log for the node just has

```
==============================================================================
Feature extraction
==============================================================================
```

and nothing else, even after leaving it for 10 minutes or so before  force rebooting the computer.

The pc has an AMD Ryzen 5 2600 cpu and an AMD rx 570 gpu. The freeze happens even if use cpy only is enabled in the node. If there are any logs that I could send please could you let me know how to get them.

Thanks!

**Comment by revisionarian** [comment]
Hi @ThePinkUnicorn6, thanks for your bug report concerning MeshroomCL freezing up during StructureFromMotionCL processing.

Based on the log message you provided, it does seem that the freeze is happening at the beginning of the feature extraction phase. Feature extraction does not use the GPU, but can heavily utilize the CPU and RAM. The current version of MeshroomCL will attempt to spawn 8 separate threads to extract features in up to 8 input images simultaneously. One possibility is that you are running out of RAM.

A few questions:
 
1. How much RAM do you have on your system?
2. How big is your input dataset (number of images and image resolution)?
3. Can you watch the RAM and CPU utilization when you start MeshroomCL processing?
4. If you run MeshroomCL on an input dataset with only 3 or 4 images, does it still freeze?

**Comment by ThePinkUnicorn6** [comment]
You are correct about it being the ram usage, I have 8 gigabytes of ram and when it starts processing it jumps to 100% used. The dataset is 60 images at 4656x3496. When I try running it with just 4 of those images it still fills up all of the ram and freezes. When using just 2 images it peaks at 6gb and is able to move on to MultiviewStereoCL where it fails, presumably from the lack of images. 

I have now downscaled all of the images to 1023x768 and it only uses 70% of the ram when using all of the images so isn't crashing and has successfully completed the modle. It is odd how it can't handle the higher resolutions bit at least it is working now.

Thanks

**Comment by revisionarian** [comment]
Thanks a lot for helping us identify and solve this issue, @ThePinkUnicorn6.

We will prepare a new release of MeshroomCL that adaptively spawns the feature extraction threads, based on the amount of available memory. This should allow you to run the software with your full resolution images.

**Comment by revisionarian** [comment]
Hi @ThePinkUnicorn6, today we have released MeshroomCL version 0.6.0, available here:

[https://github.com/openphotogrammetry/meshroomcl/releases](https://github.com/openphotogrammetry/meshroomcl/releases)

This new version of the software addresses the memory (RAM) problem that you brought to our attention, and you should now be able to process your dataset with the full resolution photos with the default settings. Additionally, we added a "Max CPU Threads" option to the StructureFromMotionCL node, which you can change to a lower value if you want to use even less memory (fewer threads will use less memory, but will require more time for processing).

Please let us know if this does not solve your problem, and don't hesitate to contact us if you have other questions or ideas regarding the software.

**Comment by ThePinkUnicorn6** [comment]
Ok thank you! I will try it out later.
